### PR TITLE
[Model] Add support for DBRX

### LINKED ISSUE #3658: [New Model]: Supporting DBRX from Databricks
### The model to consider.

Databricks has released [DBRX](https://www.databricks.com/blog/introducing-dbrx-new-state-art-open-llm), which consists of 2 models

- [dbrx](https://huggingface.co/databricks/dbrx-base)
- [dbrx-instruct](https://huggingface.co/databricks/dbrx-instruct)

It's a 132B parameter MoE model. Might be useful.

### The closest model vllm already supports.

_No response_

### What's your difficulty of supporting the model you want?

It seems that they have a custom script in their files, might need custom implementation on that regard.

### PR DESCRIPTION
**Pull request by megha95**
This PR adds supports for DBRX. DBRX is a Mixture-of-Experts (MoE) model trained by Databricks with 132B total parameters and 36B live parameters. More details about the model can be found at: [DBRX Technical Blog](https://www.databricks.com/blog/introducing-dbrx-new-state-art-open-llm)

Model weights can be found in HF repo:

| Model         | Link                                                      | Description                               |
|---------------|-----------------------------------------------------------|-------------------------------------------|
| DBRX Base     | [HF](https://huggingface.co/databricks/dbrx-base)         | Pre-trained Base Model                    |
| DBRX Instruct |  [HF](https://huggingface.co/databricks/dbrx-instruct)    | Finetuned model for instruction following |

This PR is currently based off of an [older commit](https://github.com/vllm-project/vllm/pull/3299) because the latest main has some [issues](https://github.com/vllm-project/vllm/issues/3644). Therefore, there are some minor merge conflicts that will be corrected soon.
Meanwhile, to run DBRX with vLLM, this PR can be used. It has been tested on NVIDIA A100 and H100 systems.

Note: Given model has 132B total parameters, it is suggested to use mininum 4x80GB GPU cards to run 16-bit inference. Try increasing `gpu_memory_utilization` if you are running on 4 GPUs.

### CODE REVIEW

File: vllm/transformers_utils/configs/__init__.py [hunk 1]
```diff
@@ -1,14 +1,17 @@
 from vllm.transformers_utils.configs.chatglm import ChatGLMConfig
 from vllm.transformers_utils.configs.mpt import MPTConfig
+
 # RWConfig is for the original tiiuae/falcon-40b(-instruct) and
 # tiiuae/falcon-7b(-instruct) models. Newer Falcon models will use the
 # `FalconConfig` class from the official HuggingFace transformers library.
 from vllm.transformers_utils.configs.falcon import RWConfig
 from vllm.transformers_utils.configs.starcoder2 import Starcoder2Config
+from vllm.transformers_utils.configs.dbrx import DbrxConfig
```
Comment (WoosukKwon) [inline vllm/transformers_utils/configs/__init__.py:9]: nit: Please sort these import statements in alphabetic order

File: vllm/transformers_utils/config.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
     "RefinedWeb": RWConfig,  # For tiiuae/falcon-40b(-instruct)
     "RefinedWebModel": RWConfig,  # For tiiuae/falcon-7b(-instruct)
     "starcoder2": Starcoder2Config,
+    "dbrx": DbrxConfig,
```
Comment (WoosukKwon) [inline vllm/transformers_utils/config.py:13]: nit: Please keep the alphabetic order.

File: vllm/model_executor/models/dbrx.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+        param_data = param.data
+        shard_size = self.intermediate_size
+        shard = slice(tp_rank * shard_size, (tp_rank + 1) * shard_size)
+        if weight_name.endswith("w1"):
+            loaded_weight = torch.reshape(
+                loaded_weight,
+                [-1, self.intermediate_size * self.tp_size, self.d_model],
+            )
+            param_data[:, 0:shard_size, :] = loaded_weight[:, shard, :]
+        if weight_name.endswith("v1"):
```
Comment (WoosukKwon) [inline vllm/model_executor/models/dbrx.py:143]: what does `v1` mean? Can you comment on this?

File: vllm/transformers_utils/configs/dbrx.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+"""Dbrx configuration."""
+
+from typing import Any, Optional
+
+from transformers.configuration_utils import PretrainedConfig
+from transformers.utils import logging
+
+logger = logging.get_logger(__name__)
+
+DBRX_PRETRAINED_CONFIG_ARCHIVE_MAP = {}
```
Comment (WoosukKwon) [inline vllm/transformers_utils/configs/dbrx.py:13]: Do we need this?

File: vllm/model_executor/models/dbrx.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+from vllm.model_executor.weight_utils import (
+    default_weight_loader,
+    hf_model_weights_iterator,
+)
+from vllm.sequence import SamplerOutput
+from vllm.transformers_utils.configs.dbrx import DbrxConfig
+from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.utils import set_weight_attrs
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
```
Comment (WoosukKwon) [inline vllm/model_executor/models/dbrx.py:40]: Please change `KVCache` to `torch.Tensor`. We recently changed the type :)

**Review by WoosukKwon — COMMENTED** [review]
@megha95 Thanks for submitting the PR! Very excited about the new release. Left minor comments on some stylistic issues. PTAL.

File: vllm/transformers_utils/configs/dbrx.py (hunk 4 above)
Comment (megha95) [reply vllm/transformers_utils/configs/dbrx.py:13]: Was following the same format as HF `configuration_dbrx.py` (https://huggingface.co/databricks/dbrx-base/blob/4a0342461bb367b6f9880733d3c1160d640f2942/configuration_dbrx.py#L9)

Comment (WoosukKwon) [reply vllm/transformers_utils/configs/dbrx.py:13]: Got it. Thanks for the explanation.

File: vllm/model_executor/models/dbrx.py (hunk 3 above)
Comment (megha95) [reply vllm/model_executor/models/dbrx.py:143]: DBRX uses GLU (Gated Linear Units) in expert layer. GLU has 3 linears: `w1`, `v1` and `w2`. This is similar to GLU used in LLaMa2 model.

Comment (WoosukKwon) [reply vllm/model_executor/models/dbrx.py:143]: I see. Thanks for the explanation!

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for submitting the PR! Very excited to see what people will build on top of DBRX.

**Comment by RonanKMcGovern** [comment]
doesn't this need tiktoken to be installed? I guess that's not pushed yet to the docker image maybe? could the docker image be re-built?

**Comment by Calvinnncy97** [comment]
Hi guys, is there an example command that can be used to deploy an API server for DBRX-Base with vLLM? I have been trying with `CUDA_VISIBLE_DEVICES=1,2,3,4 python -m vllm.entrypoints.openai.api_server --model databricks/dbrx-base --tensor-parallel-size 4 --host localhost --port 12345 --gpu-memory-utilization 0.9` but it seems to be asking for 1EB of memory with error message `torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate more than 1EB memory.`

**Comment by RonanKMcGovern** [comment]
I had it running with TGI fine and can't remember if I got it working with
vLLM, but I did prepare this template:
https://runpod.io/console/deploy?template=bi8ao1ztys&ref=jmfkcdio , which
uses:
```
--model databricks/dbrx-instruct --max-model-len 4096 --port 8000
--trust-remote-code
```
That does strike me that it's missing gpus all....

If you're hitting memory issues, maybe consider trying --dtype of half - to
see if that forces bfloat16 or float16?

BTW, for fast memory download, the latest vLLM docker image (and indeed
requirements.txt in the main repo) include hf_transfer, so if you set
"HF_HUB_ENABLE_HF_TRANSFER" you'll get much faster download via rust.

**Comment by Calvinnncy97** [comment]
I see. Makes sense if you set the --max-model-len to 4096. I was doing 32k context length. On the topic of precision, vLLM automatically uses FP16 for FP16 model, and BF16 for BF16 model if I am not wrong.

**Comment by Calvinnncy97** [comment]
Hi, update: I found out what the issue is. It is due to this warning:
```
WARNING 04-23 04:08:51 custom_all_reduce.py:52] Custom allreduce is disabled because it's not supported on more than two PCIe-only GPUs. To silence this warning, specify disable_custom_all_reduce=True explicitly.
```

Because custom allreduce is disabled, the model did not manage to load

**Comment by RonanKMcGovern** [comment]
ok, wow thanks, that's pretty specific

**Pull request merged**